platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/recovery/internal/EventStreamBound.java

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.recovery.internal;
+
+import com.swirlds.common.system.events.ConsensusData;
+import com.swirlds.common.system.events.DetailedConsensusEvent;
+import com.swirlds.platform.internal.EventImpl;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Instant;
+import java.util.Objects;
+
+/**
+ * A bound on the event stream. This bound can be either a round or a timestamp but not both. The bound is inclusive of
+ * exact matches to round and timestamp. If no round or timestamp is specified, then the bound is unbounded.
+ */
+public class EventStreamBound {
```

Comment:
I'm curious your thoughts on the following idea:

- create an interface `EventStreamBound`
- create 3 separate classes which implement `EventStreamBound`
    - `RoundBound`
    - `TimeBound`
    - `UnboundedBound`

- Each of these would implement the various comparison methods, and be generally quite simple.
- A builder would no longer be necessary, and you could skip the whole "make sure both things aren't set simultaneously" logic.
- If it ever becomes necessary, you could implement a `MultiBound`, that checks both round and time.

This strikes me as simpler and more understandable than the current configuration.

Reply:
I bet the net number of lines required for this strategy would be smaller, even though we would have to create more classes.

Reply:
The latest solution incorporates the suggestion to break everything up into an interface and implementations.